### PR TITLE
Directly mapping engine's data type to Java type for TRowSet generation

### PR DESCRIPTION
**Pull request by bowenliang123**
# :mag: Description
## Issue References 🔗

As described.

## Describe Your Solution 🔧

Previously, the TRowSetGenerator built a chain of `Engine's data type` to `Thrift data type` and then to `Java data type`.
This limits the possibility of overriding the implementation of Java data types handlers and coupling the `Engine's data type` to specific `Thrift data type`.

1. This PR directly mapping `Engine's data type` to `Java data type` focusing converting the upstream Engine result set's value to real data type. 
2.  it also decoupled the `Thrift data type` to Java data types, eg. Thrift's SMALL_INT and TINY_INT can be now ignored, and making the engines data type mapped to Java's Short data type.
3. it's now able to distinguish similar terms of data types with different meanings, like Java's `byte` or `byte array` and `tiny int` or `samll int` in engines and thrift. Only Java data types are represented in `TRowSetColumnGenerator` and `TRowSetColumnValueGenerator`

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Behavior Without This Pull Request :coffin:
No behaviour changes.

#### Behavior With This Pull Request :tada:
No behaviour changes.

#### Related Unit Tests
CI tests.

---

# Checklists
## 📝 Author Self Checklist

- [x] My code follows the [style guidelines](https://kyuubi.readthedocs.io/en/master/contributing/code/style.html) of this project
- [x] I have performed a self-review
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

## 📝 Committer Pre-Merge Checklist

- [x] Pull request title is okay.
- [x] No license issues.
- [x] Milestone correctly set?
- [x] Test coverage is ok
- [x] Assignees are selected.
- [x] Minimum number of approvals
- [x] No changes are requested

**Be nice. Be informative.**

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/AbstractTRowSetGenerator.scala [hunk 1]
```diff
@@ -16,28 +16,20 @@
  */
 
 package org.apache.kyuubi.engine.schema
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/AbstractTRowSetGenerator.scala:18]: the `schema` seems incorrect to me
```suggestion
package org.apache.kyuubi.engine.result
```

**Review by pan3793 — COMMENTED** [review]
Just some comments about naming, overall LGTM

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/AbstractTRowSetGenerator.scala [hunk 2]
```diff
@@ -16,28 +16,20 @@
...
-
-import scala.collection.JavaConverters._
+import java.util.{ArrayList => JArrayList}
 
 import org.apache.kyuubi.shaded.hive.service.rpc.thrift._
-import org.apache.kyuubi.shaded.hive.service.rpc.thrift.TTypeId._
-import org.apache.kyuubi.util.RowSetUtils.bitSetToBuffer
 
-trait AbstractTRowSetGenerator[SchemaT, RowT, ColumnT] {
+trait AbstractTRowSetGenerator[SchemaT, RowT, ColumnT]
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/AbstractTRowSetGenerator.scala]: OK. Now we have the traits `TRowSetGenerator`, `TColumnValueGenerator` and `TColumnGenerator`.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/TRowSetColumnGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        ret.add(value)
+      }
+      idx += 1
+    }
+    (ret, ByteBuffer.wrap(nulls.toByteArray))
+  }
+
+  def asBooleanColumn(rows: Seq[RowT], ordinal: Int): TColumn = {
+    val (values, nulls) = getColumnToList[java.lang.Boolean](rows, ordinal, true)
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/TRowSetColumnGenerator.scala]: Changed to `J*`.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/AbstractTRowSetGenerator.scala (hunk 1 above)
Comment (bowenliang123) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/AbstractTRowSetGenerator.scala]: Changed.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/TRowSetColumnGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    val (values, nulls) = getColumnToList[java.lang.Double](rows, ordinal, 0.toDouble)
+    TColumn.doubleVal(new TDoubleColumn(values, nulls))
+  }
+
+  def asStringColumn(
+      rows: Seq[RowT],
+      ordinal: Int,
+      defaultVal: String = "",
+      convertFunc: (RowT, Int) => String = null): TColumn = {
+    val (values, nulls) = getColumnToList[java.lang.String](rows, ordinal, defaultVal, convertFunc)
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/schema/TRowSetColumnGenerator.scala]: Yes, fixed.

**Comment by bowenliang123** [comment]
> Just some comments about naming, overall LGTM

Thanks for the reviews. All the comments have been adopted and resolved.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/FlinkTRowSetGenerator.scala [hunk 5]
```diff
@@ -62,9 +62,11 @@ class FlinkTRowSetGenerator(zoneId: ZoneId)
         asStringTColumnValue(
           row,
           ordinal,
-          TIMESTAMP_LZT_FORMATTER.format(
-            ZonedDateTime.ofInstant(getColumnAs[Instant](row, ordinal), zoneId)))
-      case t => asStringTColumnValue(row, ordinal, toHiveString((row.getField(ordinal), t)))
+          rawValue =>
+            TIMESTAMP_LZT_FORMATTER.format(ZonedDateTime.ofInstant(
+              rawValue.asInstanceOf[Instant],
+              zoneId)))
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/FlinkTRowSetGenerator.scala:68]:
```suggestion
            TIMESTAMP_LZT_FORMATTER.format(
              ZonedDateTime.ofInstant(rawValue.asInstanceOf[Instant], zoneId)))
```

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Pull request closed without merging**